### PR TITLE
Created preprocessing script for Botswana

### LINKED ISSUE #37: Develop preprocessing script for legislation from Botswana

### PR DESCRIPTION
**Pull request by Skyrolon**
Closes #37

### CODE REVIEW

File: raw_data/preprocessed/africa/botswana_english.txt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
...n under this Act, by virtue of that department having the necessary technical expertise to do so; "scoping" means consultations with concerned government departments, local authorities, non-governmental organisations, members of the public and any other interested persons, to determine the extent of the likely environmental impact of a proposed activity; "statement" means a report of the environmental impact assessment study executed in accordance with terms of reference approved by the competent authority; "strategic environmental assessment" means a process for evaluating the environmental consequences of proposed policy, plan, or programme initiatives in order to ensure that they are fully included and appropriately addressed at the earliest stage of decision making, on par with economic and social considerations; "threshold" in relation to an environmental impact assessment of an activity, means a measurable level or limit of an output resulting from the operation of a proposed activity indicative of the activity's environmental impact; and "terms of reference" means a document which details the main environmental issues which must be addressed in an environmental impact study.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/botswana_english.txt]: Please add a colon between `requires` and `"activity"`.

File: raw_data/preprocessed/africa/botswana_english.txt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+Requirements to undertake an activity
+(4) [1] No person shall undertake or implement an activity prescribed under section 3 unless the environmental impact of the proposed activity is fully taken into account in accordance with the provisions of this Act and authorisation has been issued under section [14]. [2] Any person who undertakes or implements an activity in contravention of subsection [1] commits an offence and is liable to a fine not exceeding P100000 or to a term of imprisonment not exceeding two years or to both. [3] A person who contravenes subsection [1] shall rehabilitate the area affected by the adverse environmental impact of the implemented activity. [4] In this section, "rehabilitate" means the reinstatement or restoration to a normal or functional environmental state. [5] A person who contravenes subsection [3] commits an offence and is liable to a fine not exceeding P15000 or to a term of imprisonment not exceeding 2 years or to both. [6] In the case of a continued contravention in respect of a person convicted under subsection [2] or [5], that person is liable to a further fine not exceeding P2000 for every day on which the offence continues.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/botswana_english.txt]: `section [14]` should read `section 14`. The rule, which I should have made clearer earlier (sorry) is that references to sub-article delimiters should be encapsulated in square brackets, while references to article numbers should not.

File: raw_data/preprocessed/africa/botswana_english.txt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+Ensuring issuance of authorisation
+(5) [1] A licensing authority shall, before issuing a permit, licence, consent or approval to any person who has applied to implement an activity prescribed under section 3, ensure that authorisation has been issued for the proposed activity in accordance with this Act. [2] In this section, "licensing authority" means any authority vested with the responsibility of issuing licenses for any of the activities prescribed under section [3].
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/botswana_english.txt]: `section [3]` should read `section 3`.

File: raw_data/preprocessed/africa/botswana_english.txt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
...e statement, to avert or minimise the potential adverse environmental impact, are effective and sufficient; or [b] where the competent authority is of the view that (i) the statement does not sufficiently identify the environmental impact likely to be caused, or [ii]  the prescribed mitigation measures to avert or minimise the potential adverse environmental impact, are insufficient and ineffective, afford the developer an opportunity to comply with the provisions of the Act; or [c] reject the statement, where the developer fails to satisfy the conditions under paragraph [b]. [2] In deferring or rejecting the statement, the competent authority shall furnish the applicant with a written statement of the reasons for its decision. [3] An authorisation issued under subsection [1][a] shall be in the prescribed form and shall include [a] what has been decided; [b] why the decision has been made; and [c] how the statement influenced the decision. [4] The competent authority may, in issuing an authorisation, prescribe, in writing, specific requirements for monitoring during and after implementation of the proposed activity, by the technical departments, local authorities and the developer.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/botswana_english.txt]: Please replace the two instances of `(i)` with `(ii)`.

File: scripts/legislation/africa/botswana_english.sh [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  sed -E 's/^\(([A-Za-z0-9]+)\)/[\1]/' 
+}
+
+function move_article_titles_above_article_bodies {
+  sed -E 's/^(\([0-9]+\)) ([^|]+)\|/\2\n\1/' 
+}
+
+function remove_all_text_after_last_article {
+  sed -E '/^\(32\) /,${/^\(32\)/!d}' | \
+  sed -E 's/\(32\) Articles deemed.*//'
```
Comment (danpdraper) [inline scripts/legislation/africa/botswana_english.sh]: Are you looking to delete this entire line? If so, you could write:

```
sed -E '/^\(32\) Articles deemed.*/d'
```

Also, this line should be further indented.

File: scripts/legislation/africa/botswana_english.sh [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+function amend_errors_in_headers {
+    sed -E 's/PART I Preliminary /PART I - Preliminary /' | \
```
Comment (danpdraper) [inline scripts/legislation/africa/botswana_english.sh]: This first line should not be indented to the same extent as the continuation lines.

File: scripts/legislation/africa/botswana_english.sh [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  cat "$input_file_path" | \
+    remove_all_text_before_first_header | \
+    prefix_article_numbers_with_article_literal | \
+    append_pipe_to_article_title | \
+    replace_parentheses_around_article_delimiters_with_square_brackets | \
+    apply_common_transformations_to_stdin "$language" | \
+    remove_all_text_after_last_article | \
+    amend_errors_in_headers | \
+    amend_errors_in_articles | \
+    move_article_titles_above_article_bodies
```
Comment (danpdraper) [inline scripts/legislation/africa/botswana_english.sh:101]: Please see my comment on the Namibia pull request regarding the arrangement of function definitions.

**Review by danpdraper — CHANGES_REQUESTED** [review]

File: raw_data/preprocessed/africa/botswana_english.txt (hunk 4 above)
Comment (Skyrolon) [reply raw_data/preprocessed/africa/botswana_english.txt]: Do you mean replace (i) with [i]?

File: scripts/legislation/africa/botswana_english.sh (hunk 5 above)
Comment (Skyrolon) [reply scripts/legislation/africa/botswana_english.sh]: This function does not remove everything as intended. This piece of legislation is a bit unique. There were two separate Acts in the pdf provided, one on EIA and one on postal services. Therefore there were two article 32. After running the function on line 25 this is the output:

> Act to bind State
> (32) This Act binds the State.
> Articles deemed to be posted in contravention of Act
> (32) A postal article received in a post office shall be deemed

I wasn't sure how to locate the first article 32 and delete all lines after it so this was my solution. Please let me know if there is a better way to do it.

File: raw_data/preprocessed/africa/botswana_english.txt (hunk 4 above)
Comment (danpdraper) [reply raw_data/preprocessed/africa/botswana_english.txt]: Oops, yes, that is what I meant. Please replace `(i)` with `[i]`. Sorry about that.

File: scripts/legislation/africa/botswana_english.sh (hunk 5 above)
Comment (danpdraper) [reply scripts/legislation/africa/botswana_english.sh]: I figured that was what you were doing and I agree with your approach in general. I am just saying that you can modify the second of the two lines in the function as suggested. The complete function would be written as follows:

```
function remove_all_text_after_last_article {
  sed -E '/^\(32\) /,${/^\(32\)/!d}' | \
    sed -E '/^\(32\) Articles deemed.*/d'
}
```

My original comment only concerned the second of the two lines in the function.

Comment (Skyrolon) [reply scripts/legislation/africa/botswana_english.sh]: Thanks for the clarification!

**Review by danpdraper — APPROVED** [review]

**Pull request merged**